### PR TITLE
Exchange

### PR DESCRIPTION
**Pull request by frostornge**
#15

### CODE REVIEW

File: exchange/exchange.go [hunk 1]
```diff
@@ -35,30 +48,68 @@ func (exchange *Exchange) Order(ctx context.Context, offeror, offeree, contract
...
 
-	event, err := exchange.contract.ParseOfferPresentedFromReceipt(receipt)
+	event, err := exchange.contract.ParseOfferPreparedFromReceipt(receipt)
 	if err != nil {
 		return ablCommon.ID{}, errors.Wrap(err, "failed to parse event from the receipt")
 	}
+	offerId := ablCommon.ID(event.OfferId)
+
+	if ids == nil {
+		l := len(dataIds)
```
Comment (therne) [inline exchange/exchange.go]: plz add comment

File: contracts/contracts/DataRegistry.sol [hunk 2]
```diff
@@ -56,13 +57,10 @@ contract DataRegistry is Ownable {
             "You have been allowed to collect the data at that time. Why is it a problem?"
         );
 
-        bool challengeIsTrue = sparseMerkleTree.checkMembership(bundle.usersRoot, uint64(userId), proof);
```
Comment (therne) [inline contracts/contracts/DataRegistry.sol]: former code is better

File: exchange/exchange.go (hunk 1 above)
Comment (frostornge) [reply exchange/exchange.go]: k, i will do this on next PR

File: contracts/contracts/DataRegistry.sol (hunk 2 above)
Comment (frostornge) [reply contracts/contracts/DataRegistry.sol]: agree

**Pull request merged**